### ISSUE TITLE
Extend sendMail so that it can return the text and/or html content of the email 

### ISSUE THREAD

**Issue by tapirus** [cfwheels/cfwheels#575]
Extend sendMail so that it can return the text and/or html content of the email.

It would be useful if you would like to log the content of sent mails for example.

More info here: https://groups.google.com/forum/#!topic/cfwheels/ukiCFove04E

**Comment by perdjurner** [comment]
This is already possible according to a post on that topic.

Can I close this issue?

**Comment by andybellenie** [comment]
Is it documented?

**Comment by chapmandu** [comment]
It's not documented but it's a matter of passing in a $deliver=false argument.. 

I think this is a very worthwhile update as I use it for previewing an email's html including layout before sending..

**Comment by tapirus** [comment]
Hi, this is not possible. Only for not sent test emails.Therefore I prefer extending wheels and not closing this ticket.Thanks Thorsten
Here is the thread:Google Groups
|   |
|   |   |   |   |   |
| Google GroupsGoogle Groups allows you to create and participate in online forums and email-based groups with a rich experience for community conversations. |
|  |
| Auf groups.google.com anzeigen | Vorschau nach Yahoo |
|  |
|   |


 

    Adam Chapman <notifications@github.com> schrieb am 7:40 Dienstag, 2.Februar 2016:
 

 It's not documented but it's a matter of passing in a $deliver=false argument.. I think this is a very worthwhile update as I use it for previewing an email's html including layout before sending.. —

**Comment by andybellenie** [comment]
Lets make it a part of the public api then

**Comment by chapmandu** [comment]
I'd like to see these additional arguments:

deliver=false - executes/skips cfmail
file="/somepath/somefile.txt" - writes the generated string to a file (either full path or relative to /files)
output="someVarName" - returns generated string to variable

**Comment by chapmandu** [comment]
I've been playing around with the sendEmail method and need some opinions.

Currently the function returns null. I propose returning the tagContent string if it's not a multipart email. 

If it is a multipart email I can either return an array of tagContent strings or a struct returning each string using its type as the key eg: `{text="foo", html="bar"}`

I would have like to be able to use the `file="somefile.txt"` to write the email output to a file but alas it is already being used for attachments.. Would it be too cheeky to change the current `file` argument to `attachment` and re-purpose `file`? Anyway, I am current working argument is `destination`.

I've also made the `deliver` argument part of the api.

// don't send but write to file
sendEmail(deliver=false, destination=myPath, ...);

// send the email and save email body to a variable
emailContent = sendEmail(...);

// send a multipart email, save email body to a variable and write to a file
emailContent = sendEmail(
        templates="/myemailtemplateplain,/myemailtemplatehtml",
        layouts="customerPlain,customerHtml",
        destination=myPath
);
// dump the array/struct
dump(emailContent);

In the case of writing a multipart email to a file, both email bodies will be written separated by a crlf.

So in summary:

1. Return email body as a string, array or struct?
2. Should the `file/s` argument be repurposed and replaced with `attachment/s`
3. Should `destination` be used to write body to a file
4. `deliver` has been made part of the api

Thoughts?

**Comment by neokoenig** [comment]
Re-purposing `file`, whilst it makes sense (and I think `attachments` is a better key name to use) would break existing code; so really, it's more of a question of 'is it worth it' :)

The other argument is that, if you can now set the response as a variable/struct, there's nothing stopping you adding a `cflog/cffile` command after the `sendEmail()` call: so really, you're adding a convenience method at the detriment of existing code: from that perspective, I'd say it isn't worth it.

`destination` risks being confusing too maybe? How about something much more explicit? like `writeToFile` or something? That at least tells you what the argument will achieve.

Re: whether to return struct/string etc; I'd expect something like `html` and `plain` to always be returned: wheels should already know if the content is html or not, so you should be able to populate the returned struct appropriately. But I'd always return both struct keys, even if one is just empty. Having it as a struct also means that if in the future we want to return more information from the `sendEmail` method (I dunno, like whether the email was spooled or not, or a mailerid or something), we can just add another struct key easily; with a string it would be more likely to break the then existing code again.

Re: `deliver=false`, great, I've no issue with that at all.

**Comment by perdjurner** [comment]
Was there something wrong with returning all cfmail arguments like it used to (when $deliver=false I mean)?

**Comment by chapmandu** [comment]
It will still return all cfmail arguments, but will now also return the generated email content.

**Comment by tapirus** [comment]
This is great news! Thanks! Thorsten 

    Adam Chapman <notifications@github.com> schrieb am 21:58 Dienstag, 22.März 2016:
 

 It will still return all cfmail arguments, but will now also return the generated email content. —

**Comment by neokoenig** [comment]
Closing, new issue added to document.
